DirectProgramming/DPC++FPGA/Tutorials/DesignPatterns/zero_copy_data_transfer/src/restricted_usm_kernel.hpp

```diff
@@ -0,0 +1,142 @@
+// Decoupling the memory accesses and using SYCL pipes with a substantial
+// depth ('kPipeDepth' below) allows the kernel to be more resilient against
+// stalls while reading/writing from/to the Host Memory over PCIe. 
+//
+
+// Declare the kernel class names globally to avoid name mangling.
+class RestrictedUSM;
+class Producer;
+class Consumer;
+
+// pipes
+constexpr size_t kPipeDepth = 64;
+template <typename T>
+using ProducePipe = pipe<class ProducePipeClass, T, kPipeDepth>;
+template <typename T>
+using ConsumePipe = pipe<class ConsumePipeClass, T, kPipeDepth>;
+
+//
+// reads the input data from the hosts memory
+// and writes it to the ProducePipe
+//
+template <typename T>
+event SubmitProducer(queue& q, T* in_data, size_t size) {
+  auto e = q.submit([&](handler& h) {
+    h.single_task<Producer>([=]() [[intel::kernel_args_restrict]] {
+      // using a host_ptr tells the compiler that this pointer lives in the
+      // hosts address space
+      host_ptr<T> h_in_data(in_data);
+
+      for (size_t i = 0; i < size; i++) {
+        const T data_from_host_memory = *(h_in_data + i);
+        ProducePipe<T>::write(data_from_host_memory);
+      }
+    });
+  });
+
+  return e;
+}
+
+//
+// The worker kernel in the device:
+//  1) read input data from the ProducePipe
+//  2) perform computation
+//  3) write the output data to the ConsumePipe
+//
+template <typename T>
+event SubmitWorker(queue& q, size_t size) {
+  auto e = q.submit([&](handler& h) {
+    h.single_task<RestrictedUSM>([=]() [[intel::kernel_args_restrict]] {
+      for (size_t i = 0; i < size; i++) {
+        const T data = ProducePipe<T>::read();
+        const T value = data * i; // perform computation
+        ConsumePipe<T>::write(value);
+      }
+    });
+  });
+
+  return e;
+}
+
+//
+// reads output data from the device via ConsumePipe
+// and writes it to the hosts memory
+//
+template <typename T>
+event SubmitConsumer(queue& q, T* out_data, size_t size) {
+  auto e = q.submit([&](handler& h) {
+    // using a host_ptr tells the compiler that this pointer lives in the
+    // hosts address space
+    host_ptr<T> h_out_data(out_data);
+
+    h.single_task<Consumer>([=]() [[intel::kernel_args_restrict]] {
+      for (size_t i = 0; i < size; i++) {
+        const T data_to_host_memory = ConsumePipe<T>::read();
+        *(h_out_data + i) = data_to_host_memory;
+      }
+    });
+  });
+
+  return e;
+}
+
+template <typename T>
+double RestrictedUSMKernel(queue& q, T* in, T* out, size_t size) {
+  // start the timer
+  auto start = high_resolution_clock::now();
+
+  // start the kernels
+  auto worker_event = SubmitWorker<T>(q, size);
+  auto producer_event = SubmitProducer<T>(q, in, size);
+  auto consumer_event = SubmitConsumer<T>(q, out, size);
+
+  // wait for all the kernels to finish
+  q.wait();
+
+  // stop the timer
+  auto end = high_resolution_clock::now();
+  duration<double, std::milli> diff = end - start;
+
+  return diff.count();
```

Comment:
Fuse in 1 line?

Reply:
I like being explicit here. Just my opinion :)